examples/oauth.rs

```diff
@@ -0,0 +1,261 @@
+    oauth_client: Extension<BasicClient>,
+) -> impl IntoResponse {
+    // Get an auth token
+    let token = oauth_client
+        .exchange_code(AuthorizationCode::new(query.code.clone()))
+        .request_async(async_http_client)
+        .await
+        .unwrap();
+
+    // Fetch user data from discord
+    let client = reqwest::Client::new();
+    let user_data: User = client
+        // https://discord.com/developers/docs/resources/user#get-current-user
+        .get("https://discordapp.com/api/users/@me")
+        .bearer_auth(token.access_token().secret())
+        .send()
+        .await
+        .unwrap()
+        .json::<User>()
+        .await
+        .unwrap();
+
+    // Create a new session filled with user data
+    let mut session = Session::new();
+    session.insert("user", &user_data).unwrap();
+
+    // Store session and get corresponding cookie
+    let cookie = store.0.store_session(session).await.unwrap().unwrap();
+
+    let cookie = Cookie::build(COOKIE_NAME, cookie)
+        .same_site(SameSite::Lax)
+        .path("/")
+        .finish();
+
+    // Set cookie
+    let r = http::Response::builder()
+        .header("Location", "/")
+        .header(SET_COOKIE, cookie.to_string())
+        .status(302);
+
+    r.body(Body::empty()).unwrap()
+}
+
+// Utility to save some lines of code
+pub struct Redirect(String);
+impl<'a> IntoResponse for Redirect {
+    fn into_response(self) -> http::Response<Body> {
+        let builder = http::Response::builder()
+            .header("Location", self.0)
+            .status(StatusCode::FOUND);
+        builder.body(Body::empty()).unwrap()
+    }
+}
+
+pub struct AuthRedirect;
+impl IntoResponse for AuthRedirect {
+    fn into_response(self) -> http::Response<Body> {
+        Redirect("/auth/discord".to_string()).into_response()
+    }
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for User
+where
+    B: Send,
+{
+    // If anything goes wrong or no session is found, redirect to the auth page
+    type Rejection = AuthRedirect;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let extract::Extension(store) = extract::Extension::<MemoryStore>::from_request(req)
+            .await
+            .expect("`MemoryStore` extension is missing");
+
+        let cookies = extract::TypedHeader::<headers::Cookie>::from_request(req)
+            .await
+            .expect("could not get cookies");
+
+        let session_cookie = match cookies.get(COOKIE_NAME) {
+            Some(c) => c,
+            None => return Err(AuthRedirect),
+        };
+
+        let session = match store
+            .load_session(session_cookie.to_string())
+            .await
+            .unwrap()
+        {
+            Some(c) => c,
+            None => return Err(AuthRedirect),
+        };
+
+        let user = match session.get::<User>("user") {
+            Some(u) => u,
+            None => return Err(AuthRedirect),
+        };
```

Comment:
These three `match`es could also be written using `.ok_or(AuthRedirect)?` which is a bit cleaner imo.